compiler/p/codegen/OMRCodeGenerator.cpp

```diff
@@ -3168,7 +3168,7 @@ OMR::Power::CodeGenerator::loadIntConstantFixed(
       cursor = self()->getAppendInstruction();
       }
 
-   cursor = firstInstruction = generateTrg1ImmInstruction(self(), TR::InstOpCode::lis, node, trgReg, value>>16, cursor);
+   cursor = firstInstruction = generateTrg1ImmInstruction(self(), TR::InstOpCode::lis, node, trgReg, (int16_t)(value>>16), cursor);
```

Comment:
The cast here is unnecessary, as a 32-bit value algebraically shifted right by 16 places is already guaranteed to be in the range of a 16-bit signed integer.

Reply:
To be honest, I am not sure if it's implementation defined or not.

Reply:
Technically, this is implementation-defined until C++20 (which now mandates arithmetic shifts [1]), but all of our build compilers use arithmetic shifts and we rely on that behaviour in many places throughout the codebase. As this is not undefined behaviour, it's perfectly acceptable to rely on this behaviour.

[1] https://en.cppreference.com/w/cpp/language/operator_arithmetic#Bitwise_shift_operators

Reply:
Can we guarantee that all compilers used to build omr have the same behaviour?

Reply:
Yes. All compilers that can build OMR have this behaviour and is widely accepted as the expected behaviour, hence why it was later made part of the standard. Very basic things would not work correctly if a compiler that did not have this behaviour was used to build OMR, including our own arithmetic right shift simplifier [1] and our ability to generate `slwi`/`sldi` instructions [2] [3].

[1] https://github.com/eclipse/omr/blob/24be035c96b93fae74f5396c0c7ebd75ded26cfe/compiler/optimizer/OMRSimplifierHandlers.cpp#L10350
[2] https://github.com/eclipse/omr/blob/24be035c96b93fae74f5396c0c7ebd75ded26cfe/compiler/p/codegen/GenerateInstructions.cpp#L537
[3] https://github.com/eclipse/omr/blob/24be035c96b93fae74f5396c0c7ebd75ded26cfe/compiler/p/codegen/GenerateInstructions.cpp#L563